### PR TITLE
CUDA BFloat16 TopK

### PR DESCRIPTION
**Pull request by zasdfgbnm**

### CODE REVIEW

File: aten/src/THC/THCDeviceUtils.cuh [hunk 1]
```diff
@@ -39,6 +41,15 @@ __device__ __forceinline__ T doLdg(const T* p) {
 #endif
 }
 
+template <>
+__device__ __forceinline__ c10::BFloat16 doLdg<c10::BFloat16>(const c10::BFloat16* p) {
+#if __CUDA_ARCH__ >= 350
```
Comment (ngimel) [inline aten/src/THC/THCDeviceUtils.cuh:46]: do you need #if here? torch is only supported on __CUDA_ARCH__ >= 350

Comment (zasdfgbnm) [reply aten/src/THC/THCDeviceUtils.cuh:46]: I think this is actually equivalent to `#ifndef __HIP_PLATFORM_HCC__`?

Comment (ngimel) [reply aten/src/THC/THCDeviceUtils.cuh:46]: Then it should say so? `__ldg` doesn't provide performance benefit these days, but I guess you still need to load short and construct bfloat16 from bits on cuda, and hip is able to handle it natively?

Comment (zasdfgbnm) [reply aten/src/THC/THCDeviceUtils.cuh:46]: Yes, I will change this to `#ifndef __HIP_PLATFORM_HCC__`, and maybe remove it later (needs benchmark). On HIP, it is just `*p`, so it's OK.

Comment (zasdfgbnm) [reply aten/src/THC/THCDeviceUtils.cuh]: fixed

**Comment by mcarilli** [comment]
if you observe weird numerical behavior with bfloat16 topk, https://github.com/pytorch/pytorch/blame/b85568a54a9c60986235ad1e0cc5dffc71b9d5b1/aten/src/ATen/native/cuda/SortingRadixSelect.cuh#L147-L163 is the main suspect.  @ngimel you remember our adventures with that for fp16. The same fix was also necessary for bfloat16, and @gchanan included the fix for bfloat16 in his PR, but we had no way to test bfloat16 on cuda at the time.

**Comment by zasdfgbnm** [comment]
@mcarilli Tests on CI are passing, so it should be OK? Do you think we need more tests beyond the existing unit tests?

**Review by ngimel — APPROVED** [review]

**Comment by ngimel** [comment]
CI is enough, get_all_dtypes is testing bfloat16, right?

**Comment by zasdfgbnm** [comment]
@ngimel Yes, by default it include everything, unless you say `include_bfloat=False`:
```
>>> torch.testing.get_all_dtypes()
[torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64, torch.float32, torch.float64, torch.float16, torch.bfloat16, torch.bool, torch.complex64, torch.complex128]
>>> torch.testing.get_all_fp_dtypes()
[torch.float32, torch.float64, torch.float16, torch.bfloat16]

```

**Comment by ngimel** [comment]
There are internal build failures
```
stderr: caffe2/aten/src/THC/THCDeviceUtils.cuh(47): error: calling a constexpr __host__ function("from_bits") from a __device__ function("doLdg") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
caffe2/c10/util/TypeCast.h(27): warning: calling a constexpr __host__ function("real") from a __host__ __device__ function("apply") is not allowed. The experimental flag '--expt-relaxed-constexpr' can be used to allow this.
          detected during:
            instantiation of "decltype(auto) c10::maybe_real<true, src_t>::apply(src_t) [with src_t=c10::complex<double>]" 
(57): here
            instantiation of "uint8_t c10::static_cast_with_inter_type<uint8_t, src_t>::apply(src_t) [with src_t=c10::complex<double>]" 
(157): here
            instantiation of "To c10::convert<To,From>(From) [with To=uint8_t, From=c10::complex<double>]" 
(169): here
            instantiation of "To c10::checked_convert<To,From>(From, const char *) [with To=uint8_t, From=c10::complex<double>]" 
```

**Comment by zasdfgbnm** [comment]
Let me benchmark and remove ldg

**Comment by ngimel** [comment]
The problem is not __ldg, I believe, it's fromBits. I have no idea why --expt-relaxed-constexpr is not passed in internal builds and how it used to work. Maybe just returning CUDA_ARCH guard is the way to go ;-)

**Comment by zasdfgbnm** [comment]
The solution for the `__ldg` is at https://github.com/pytorch/pytorch/pull/44925, I will rebase and fix this after that PR is merged.

File: aten/src/THC/THCDeviceUtils.cuh (hunk 1 above)
Comment (nickbeckwith) [reply aten/src/THC/THCDeviceUtils.cuh]: Just to let you know, changing the #if does break the build for NVIDIA GRID K520 GPU.  I understand that is not a supported CUDA architecture.

Comment (ngimel) [reply aten/src/THC/THCDeviceUtils.cuh]: Sorry about that, but as you note it is not a supported architecture.

**Pull request closed without merging**